pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -0,0 +1,865 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"bytes"
+	"net/netip"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+
+	"golang.org/x/exp/maps"
+	"google.golang.org/protobuf/proto"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	klabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/api/security/v1beta1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/istio/pkg/config/schema/gvk"
+	"istio.io/istio/pkg/config/schema/kind"
+	"istio.io/istio/pkg/kube/controllers"
+	kubelabels "istio.io/istio/pkg/kube/labels"
+	"istio.io/istio/pkg/spiffe"
+	"istio.io/istio/pkg/util/sets"
+	"istio.io/istio/pkg/workloadapi"
+)
+
+// AmbientIndex maintains an index of ambient WorkloadInfo objects by various keys.
+// These are intentionally pre-computed based on events such that lookups are efficient.
+type AmbientIndex struct {
+	mu sync.RWMutex
+	// byService indexes by Service (virtual) *IP address*. A given Service may have multiple IPs, thus
+	// multiple entries in the map. A given IP can have many workloads associated.
+	byService map[string][]*model.WorkloadInfo
+	// byPod indexes by Pod IP address.
+	byPod map[string]*model.WorkloadInfo
+
+	// Map of ServiceAccount -> IP
+	waypoints map[types.NamespacedName]sets.String
+}
+
+// Lookup finds a given IP address.
+func (a *AmbientIndex) Lookup(ip string) []*model.WorkloadInfo {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	// First look at pod...
+	if p, f := a.byPod[ip]; f {
+		return []*model.WorkloadInfo{p}
+	}
+	// Fallback to service. Note: these IP ranges should be non-overlapping
+	return a.byService[ip]
+}
+
+func (a *AmbientIndex) dropWorkloadFromService(svcAddress, workloadAddress string) {
+	wls := a.byService[svcAddress]
+	// TODO: this is inefficient, but basically we are trying to update a keyed element in a list
+	// Probably we want a Map? But the list is nice for fast lookups
+	filtered := make([]*model.WorkloadInfo, 0, len(wls))
```

Comment:
Could find the index of the item to delete, swap it to the end and then return `[:len(wls) - 1`

Reply:
Good point. I would note this is just moved code though